physioqc/metrics/cardiac.py

```diff
@@ -1,8 +1,564 @@
 """Denoising metrics for cardio recordings."""
 
+import matplotlib.pyplot as plt
 import numpy as np
+from matplotlib.patches import Polygon
+from numpy.polynomial import Polynomial
+from peakdet import operations
+from scipy.signal import resample, welch
+from scipy.stats import kurtosis, pearsonr, skew
 
 from .. import references
 from ..due import due
+from .utils import butterbpfiltfilt, butterlpfiltfilt, hamming, physio_or_numpy
 
-pass
+
+def cardiacprefilter(rawcard, Fs, lowerpass=0.1, upperpass=10.0, order=1, debug=False):
+    if debug:
+        print(
+            f"cardiacprefilter: Fs={Fs} order={order}, lowerpass={lowerpass}, upperpass={upperpass}"
+        )
+    return butterbpfiltfilt(Fs, lowerpass, upperpass, rawcard, order, debug=debug)
+
+
+def cardenvelopefilter(squarevals, Fs, upperpass=0.25, order=8, debug=False):
+    if debug:
+        print(f"cardenvelopefilter: Fs={Fs} order={order}, upperpass={upperpass}")
+    return butterlpfiltfilt(Fs, upperpass, squarevals, order, debug=debug)
+
+
+def trendgen(thexvals, thefitcoffs, demean):
+    """
+
+    Parameters
+    ----------
+    thexvals: array-like
+        The x-axis value at which to calculate the trend
+    thefitcoffs: array-like
+        The polynomial coefficients used to calculate the trend
+    demean: bool
+        Flag to determine if data should be demeaned
+
+    Returns
+    -------
+
+    """
+    theshape = thefitcoffs.shape
+    order = theshape[0] - 1
+    thepoly = thexvals
+    thefit = 0.0 * thexvals
+    if order > 0:
+        for i in range(1, order + 1):
+            thefit += thefitcoffs[order - i] * thepoly
+            thepoly = np.multiply(thepoly, thexvals)
+    if demean:
+        thefit = thefit + thefitcoffs[order]
+    return thefit
+
+
+def detrend(inputdata, order=1, demean=False):
```

Comment:
@smoia What do you think about the idea of integrating that function in the peakdet operations module instead ?

Reply:
Detrend?  Sure.  When I went to fork peakdet, I saw that the description says "Former toolbox for physiological peak detection analyses. Deprecated in favour of prep4phys".  So should I be using that instead?

Reply:
That's a good question ! @smoia @m-miedema  WDYT ?

Reply:
I would say let's keep working on peakdet and we'll project the changes to prep4phys. I'm surprised peakdet didn't have detrending options before!